### PR TITLE
fix: whenCurrentJobsFinished shouldn't initialize bclient

### LINKED ISSUE #1346: whenCurrentJobsFinished initializes bclient unnecessarily
## Description

Bull typically does a good job of lazily connecting to Redis. But the following code causes the blocking client to be initialized unnecessarily:

https://github.com/OptimalBits/bull/blob/c0ee6be03c6d47d60465d2bae0d2566d5bf0d661/lib/queue.js#L1140-L1147

If `bclient` hasn't been initialized yet, then we're connecting for no reason.  

## Minimal, Working Test code to reproduce the issue.
```js
const Queue = require('bull');
const Redis = require('ioredis');

const queue = new Queue('foo', {
  createClient (type) {
    if (type === 'bclient') {
      // enforce that we don't waste connections for read-only queues
      throw new Error('This queue is read-only');
    }

    return new Redis();
  }
});
queue.whenCurrentJobsFinished();
```

## Bull version
`develop` branch

## Additional information
Also reproducible when calling `queue.pause(true)` since that calls `whenCurrentJobsFinished` internally. Workaround for that is to call `queue.pause(true, true)` so that `whenCurrentJobsFinished` is not called. But that only works if you're sure there are no jobs to wait for.

### PR DESCRIPTION
**Pull request by gabegorelick**
Fixes #1346

### CODE REVIEW

File: lib/queue.js [hunk 1]
```diff
@@ -279,6 +279,12 @@ function redisClientGetter(queue, options, initCallback) {
         return connections[type];
       }
       const client = (connections[type] = createClient(type, options.redis));
+
+      // Since connections are lazily initialized, we can't check queue.client
+      // without initializing a connection. So expose a boolean we can safely
+      // query.
+      queue[type + 'Initialized'] = true;
```
Comment (manast) [inline lib/queue.js]: Couldn't you store this information in a separate data structure? furthermore, what other queue types would benefit from storing this data?

Comment (gabegorelick) [reply lib/queue.js]: > Couldn't you store this information in a separate data structure?

Not sure I follow. Do you mean you want a dedicated object on the queue to save this? Something like
```js
queue._clientStatuses = {}; // in Queue constructor

queue._clientStatuses[type] = 'initialized'; // after initialization
```

Or are you thinking there's a more complex state machine that should be used?

> what other queue types would benefit from storing this data?

Do you mean Redis connection types? There may be instances where `client` and `eclient` are also initialized unnecessarily, but I haven't tracked them down yet. That's also blocked by #1344 which means that `client` is always initialized.

**Comment by gabegorelick** [comment]
The failing test appears to be an existing race condition:

https://github.com/OptimalBits/bull/blob/c0ee6be03c6d47d60465d2bae0d2566d5bf0d661/test/test_connection.js#L82-L117

`isReady` only waits for `client` to be ready, not the `subscriber` client. `subscriber` client is initialized in the `Queue` constructor when we call `this.on('error')`, so it's a race condition whether it's ready or not by the time the test gets to that assertion. I don't see anything in my PR that would change this.

File: lib/queue.js (hunk 1 above)
Comment (gabegorelick) [reply lib/queue.js]: @manast Have you had a chance to look at this?

Comment (manast) [reply lib/queue.js]: yes, sorry. I meant a separate data structure.

**Comment by mgreenw** [comment]
This is awesome, thanks for fixing this @gabegorelick!

**Pull request merged**